File: examples/go/integ/subscription_test.go
// Copyright (c) 2022-2026 Super Durable, Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

package integ

import (
	"testing"
	"time"

	"github.com/stretchr/testify/require"
	"github.com/superdurable/dex/examples/go/workflows"
	"github.com/superdurable/dex/examples/go/workflows/subscription"
	"github.com/superdurable/dex/sdk-go/dex"
)

func TestSubscriptionStartRPCAndChannels(t *testing.T) {
	ctx := integrationContext(t)
	flowID := newFlowID(t, "subscription")
	customer := subscription.Customer{
		FirstName: "Example",
		LastName:  "Customer",
		ID:        flowID,
		Email:     "customer@example.com",
		Subscription: subscription.Subscription{
			TrialPeriod:         30 * time.Second,
			BillingPeriod:       30 * time.Second,
			MaxBillingPeriods:   2,
			BillingPeriodCharge: 100,
		},
	}
	runID, err := integClient.StartFlow(
		ctx,
		workflows.Subscription,
		flowID,
		customer,
		dex.StartFlowOptions{},
	)
	require.NoError(t, err)
	require.NotEmpty(t, runID)
	require.NoError(t, integClient.WaitForAttributeEqual(
		ctx,
		flowID,
		subscription.BillingPeriodNumber,
		0,
		dex.WaitOptions{Timeout: 20 * time.Second},
	))

	var current subscription.Subscription
	require.NoError(t, integClient.InvokeRPC(
		ctx,
		flowID,
		workflows.Subscription.Describe,
		nil,
		&current,
		dex.InvokeOptions{},
	))
	require.Equal(t, 100, current.BillingPeriodCharge)
	require.NoError(t, integClient.PublishToChannel(
		ctx,
		flowID,
		subscription.UpdateChargeAmount,
		250,
	))
	require.Eventually(t, func() bool {
		err = integClient.InvokeRPC(
			ctx,
			flowID,
			workflows.Subscription.Describe,
			nil,
			&current,
			dex.InvokeOptions{},
		)
		return err == nil && current.BillingPeriodCharge == 250
	}, 20*time.Second, 200*time.Millisecond, "Describe failed: %v", err)
	require.NoError(t, integClient.PublishToChannel(
		ctx,
		flowID,
		subscription.CancelSubscription,
		nil,
	))

	result := waitForFlow(t, flowID)
	require.Equal(t, dex.FlowCompleted, result.Status)
	require.Len(t, result.Completions, 1)
	var output string
	require.NoError(t, result.Completions[0].Output.Decode(&output))
	require.Equal(t, "subscription canceled", output)
}

File: examples/typescript/test/integ/subscription.test.ts
/*
 * Copyright (c) 2022-2026 Super Durable, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import assert from "node:assert/strict";
import test from "node:test";

import { stringCodec } from "@superdurable/dex";

import {
  acquireIntegEnvironment,
  awaitCondition,
  releaseIntegEnvironment,
} from "./environment.js";
import type { Customer, Subscription } from "../../src/workflow/subscription/models.js";

test.before(async () => {
  await acquireIntegEnvironment();
});

test.after(async () => {
  await releaseIntegEnvironment();
});

test("subscriptionStartRpcAndChannels", async () => {
  const environment = await acquireIntegEnvironment();
  const flow = environment.subscriptionFlow;
  const flowId = environment.newFlowId("subscription");
  const customer: Customer = {
    firstName: "Example",
    lastName: "Customer",
    id: flowId,
    email: "customer@example.com",
    subscription: {
      trialPeriodMs: 30_000,
      billingPeriodMs: 30_000,
      maxBillingPeriods: 2,
      billingPeriodCharge: 100,
    },
  };

  const runId = await environment.client.startFlow(
    flow,
    flowId,
    customer,
    environment.startOptions(),
  );
  assert.ok(runId.length > 0);

  await awaitCondition(
    () => environment.client.getAttribute(flowId, flow.customerDetails),
    (details) =>
      details !== undefined &&
      details.id === flowId &&
      details.subscription.billingPeriodCharge === 100,
    20_000,
    "customer details not ready",
  );

  const current = (await environment.client.invokeRPC(flow.describe, flowId)) as Subscription;
  assert.equal(current.billingPeriodCharge, 100);

  await environment.client.publish(flowId, flow.updateChargeAmount, 250);
  await awaitCondition(
    async () => (await environment.client.invokeRPC(flow.describe, flowId)) as Subscription,
    (subscription) => subscription.billingPeriodCharge === 250,
    20_000,
    "Describe charge amount did not update",
  );

  await environment.client.publish(flowId, flow.cancelSubscription, undefined);
  const output = await environment.client.waitForFlow(flowId, 45_000).then((result) =>
    result.singleOutput(stringCodec),
  );
  assert.equal(output, "subscription canceled");
});
